kvserver: clean up and fix truncations

pav-kv opened the pull request:
This PR further iterates on the raft log truncations stack code, and makes it cleaner. It also fixes a bug in log size maintenance.

The `RaftExpectedFirstIndex == 0` case now renders in distrusting the log size delta. Example scenario where the bug would previously occur:

- leader’s log is `(20, 80]`
- follower’s log is `(50, 80]` (e.g. it was inited from a snapshot)
- leader proposes a truncation <= `70`, with `sizeDelta = -size(20,70]`
- the follower does apply it, and does `raftLogSize += delta`
- the follower’s size is inaccurate (even though it believes otherwise)

Epic: none
Release note: none

pav-kv committed:
kvserver: factor out truncation staging

This is a mechanical move of the code, no changes.

Epic: none
Release note: none

pav-kv committed:
kvserver: inline loosely coupled trunc setting read

Epic: none
Release note: none

pav-kv committed:
kvserver: move truncation monotonicity check out

This commit makes the next > prev truncated index check explicit on the
caller's end, to make the logic there explicit.

Epic: none
Release note: none

pav-kv committed:
kvserver: add explicit discarding logic

This commit makes the discarding logic even more explicit, and replaces
a mutable variable with an immutable one, for better readability.

Epic: none
Release note: none

pav-kv committed:
kvserver: fix log size stats bug

The RaftExpectedFirstIndex == 0 case should render in distrusting the
log size delta.

Example scenario where the bug would occur:

- leader’s log is (20, 80]
- follower’s log is (50, 80] (e.g. it was inited from a snapshot)
- leader proposes a truncation <= 70, with sizeDelta = -size(20,70]
- the follower does apply it, and does raftLogSize += delta
- the follower’s size is inaccurate (even though it believes otherwise)

Epic: none
Release note: none

pav-kv committed:
kvserver: use pendingTruncation

Epic: none
Release note: none

pav-kv committed:
kvserver: remove lengthy variables

Epic: none
Release note: none

pav-kv committed:
kvserver: add helper for discarding truncation

Epic: none
Release note: none

pav-kv committed:
logstore: take RaftTruncatedState by value

Epic: none
Release note: none

pav-kv committed:
kvserver: squash stageTruncation

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143249)
<!-- Reviewable:end -->

pav-kv committed:
kvserver: update comments in truncations

Enumerate bugs in truncations stack and make some comments honest.

Epic: none
Release note: none

pav-kv committed:
kvserver: make names short and non-redundant

The types of the variables already reflect the fact that these are
"truncated state" variables, so no need to reflect that in the names.

Epic: none
Release note: none

pav-kv commented:
@tbg Commit-by-commit review is recommended to be convinced. Also recommend reading `stageTruncation` func as a whole once, and checking that it now makes sense.

pav-kv committed:
kvserver: unindent the tightly coupled case

Epic: none
Release note: none

tbg commented on pkg/kv/kvserver/replica_application_result.go:
```diff
@@ -510,8 +510,7 @@ func (r *Replica) handleTruncatedStateResultRaftMuLocked(
 	// this doesn't need any special casing.
-	expectedFirstIndexWasAccurate := r.shMu.raftTruncState.Index+1 == pt.expectedFirstIndex
-	isRaftLogTruncationDeltaTrusted := pt.isDeltaTrusted && expectedFirstIndexWasAccurate
+	isDeltaTrusted := pt.isDeltaTrusted && r.shMu.raftTruncState.Index+1 == pt.expectedFirstIndex
```
here it is _almost_ better to instead mutate `pt.isDeltaTrusted` so that there aren't two identifiers that mean different things but have the right name. I'm not saying it is better. Just thinking out loud.

tbg commented on pkg/kv/kvserver/replica_app_batch.go:
:chefs_kiss: that takes care of my earlier comment.

tbg commented on pkg/kv/kvserver/replica_application_state_machine.go:
```diff
@@ -326,7 +313,7 @@ func (sm *replicaStateMachine) handleNonTrivialReplicatedEvalResult(
 	// TODO(#93248): the strongly coupled truncation code will be removed once the
 	// loosely coupled truncations are the default.
-	if truncState != nil {
+	if truncState := rResult.GetRaftTruncatedState(); truncState != nil {
```
so if we discarded the truncation from rResult earlier (in `stageSomething`), it won't show up here, right?

pav-kv replied:
Yes. The lifetime of `rResult.GetRaftTruncatedState()` ends wherever `DiscardRaftTruncation()` is called.

- For loosely coupled truncs, it ends in `stageTruncation` (because the thing is handed over to the loosely coupled machinery right away).
- For tightly coupled truncs, it ends either in `stageTruncation` (if we see that this truncation is a no-op), or here after having applied it.

I can put a comment to this extent.

tbg reviewed (approved):
Nice cleanups, thanks for the easy review.

pav-kv commented on pkg/kv/kvserver/replica_app_batch.go:
```diff
@@ -498,24 +498,24 @@ func (b *replicaAppBatch) stageTruncation(
 	//
 	// TODO(pav-kv): make the size delta untrusted on expected first index == 0.
 	apply := !looselyCoupledTruncation || res.RaftExpectedFirstIndex == 0
```
Yeah, I felt the same way. Addressed in later commits, as you noted down the line. The goal was to do things surgically, one logical step at a time.

pav-kv commented on pkg/kv/kvserver/replica_app_batch.go:
```diff
@@ -482,23 +482,22 @@ func (b *replicaAppBatch) stageTruncation(
-	// will recompute the size. This has no correctness impact, so we are not
-	// going to bother with a long-running migration.
+	// The current state of the RaftExpectedFirstIndex == 0 quirk:
+	//   - Loosely-coupled code marks the log size as untrusted which triggers an
```
Mostly right. Didn't phrase this well enough so that caused confusion:

> which will basically be ignored (marks the stats as inaccurate but doesn't truncate)

The truncation would be carried out (provided that the `truncatedState` is not regressing). But it would mark the delta as inaccurate, which would update the `shMu.raftLogSizeTrusted` to false, and eventually that causes a recomputation in the truncations queue.

> The comment could drive that home (if it's still around at the end of the commits).

The comment is removed by the end, it's more of a note for the review.

pav-kv commented on pkg/kv/kvserver/replica_app_batch.go:
```diff
@@ -482,23 +482,22 @@ func (b *replicaAppBatch) stageTruncation(
+	// The current state of the RaftExpectedFirstIndex == 0 quirk:
+	//   - Loosely-coupled code marks the log size as untrusted which triggers an
+	//     eventual size re-computation. However, this path is never taken here.
+	//   - Tightly-coupled code marks the size trusted. There is a bug here that
```
Yeah, it's not explicitly marked as trusted. Just stays trusted (which is the default). Similarly, the comment was removed.

tbg commented:
Still LGTM
